
# Iterations: 100        
Average Execution Time: 0.001598355770111084 seconds

Average Peak Memory: 0.0288818359375 MB

Average Execution Time: 0.0038525700569152833 seconds

Average Peak Memory: 0.10086318969726563 MB

# Iterations: 1,000
Average Execution Time: 0.0020528242588043213 seconds

Average Peak Memory: 0.03667719268798828 MB

Average Execution Time: 0.006160878658294678 seconds

Average Peak Memory: 0.17426569366455077 MB

# Iterations: 10,000
Average Execution Time: 0.0007655791521072388 seconds

Average Peak Memory: 0.019989265537261963 MB

Average Execution Time: 0.005421704435348511 seconds

Average Peak Memory: 0.13456009302139282 MB


In [2]:
import numpy as np
# EXECUTION TIMES AND MEMORY FOR BFS SEARCH ACROSS 100, 1000, AND 10,000 INSTANCES:
execution_times = [0.001598, 0.003852, 0.002052, 0.006160, 0.000765, 0.005421]
peak_memory = [0.028881, 0.100863, 0.036677, 0.174265, 0.019989, 0.134560]

In [3]:
# Calculate mean
mean_execution_time = np.mean(execution_times)
mean_peak_memory = np.mean(peak_memory)

print(mean_execution_time, mean_peak_memory)

0.003308 0.08253916666666666


In [4]:
# Calculate standard deviation
std_execution_time = np.std(execution_times)
std_peak_memory = np.std(peak_memory)

print(std_execution_time, std_peak_memory)

0.001994287424954922 0.05823890309296889


In [5]:
# Calculate coefficient of variation
cv_execution_time = (std_execution_time / mean_execution_time) * 100
cv_peak_memory = (std_peak_memory / mean_peak_memory) * 100

print(cv_execution_time, cv_peak_memory)


60.286802447246735 70.55911204939339


In [6]:
from scipy.stats import sem, t

# Calculate confidence intervals (95% confidence)
confidence_interval_execution_time = t.interval(0.95, len(execution_times)-1, loc=mean_execution_time, scale=sem(execution_times))
confidence_interval_peak_memory = t.interval(0.95, len(peak_memory)-1, loc=mean_peak_memory, scale=sem(peak_memory))

print(confidence_interval_execution_time, confidence_interval_peak_memory)

(0.001015368880032725, 0.005600631119967276) (0.015587773425686402, 0.14949055990764692)


In [28]:
print("Mean time falls in CI range: ",confidence_interval_execution_time[0] <= mean_execution_time <= confidence_interval_execution_time[1])
print("Mean peak memory falls in CI range: ",confidence_interval_peak_memory[0] <= mean_peak_memory <= confidence_interval_peak_memory[1])

Mean time falls in CI range:  True
Mean peak memory falls in CI range:  True


# Evaluation

Null Hypothesis: Our sample mean values of time and peak memory are representative of the population values 


For execution time, the mean is 0.003308 seconds with a standard deviation of 0.001994, indicating some variability. The coefficient of variation (CV) is relatively high at 60.29%, suggesting a notable level of variability relative to the mean. The 95% confidence interval (0.001015 to 0.005601) gives a range within which we can be reasonably confident that the true mean execution time lies.

For peak memory, the mean is 0.0825 MB with a standard deviation of 0.0582. The coefficient of variation (CV) is 70.56%, indicating a higher level of variability relative to the mean. The 95% confidence interval (0.01559 to 0.1495) provides a range within which the true mean peak memory usage is likely to fall.

As the mean falls in the confidence interval range for both cases, we can affirm that our readings are accurate


neighbor    step 6 (10,000):
    Average Execution Time: 0.016422288632392885 seconds
    Average Peak Memory: 0.03628208589553833 MB
        
neighbor step 1 (10,000):
    Average Execution Time: 0.014009305453300477 seconds
    Average Peak Memory: 0.124567760181427 MB